Teach TypeLayout the Different Types of References

gmittert opened the pull request:
As part of SR-14273, the type layout infrastructure needs to be able to be able
to differentiate between types of scalars so it knows how to release/retain
appropriately. Right now, for example, to destroy a scalar, it blindly calls
into typeInfo's irgen functions which means it's not able to generate any of
the needed information for itself.

This patch adds a field to ScalarTypeLayout to allow them to know what kind of
reference they are and strings through the machinery to provide the information
to set it.

This also moves ScalarTypeLayout::destroy to use the new information.

Commit:
Teach TypeLayout the Different Types of References

Summary:
As part of SR-14273, the type layout infrastructure needs to be able to be able
to differentiate between types of scalars so it knows how to release/retain
appropriately. Right now, for example, to destroy a scalar, it blindly calls
into typeInfo's irgen functions which means it's not able to generate any of
the needed information for itself.

This patch adds a field to ScalarTypeLayout to allow them to know what kind of
reference they are and strings through the machinery to provide the information
to set it.

This also moves ScalarTypeLayout::destroy to use the new information.

Test Plan: ninja check-swift

Reviewers: mren, #pika_compiler

Reviewed By: mren

Subscribers: apl, phabricatorlinter

Differential Revision: https://phabricator.intern.facebook.com/D30983093

Tasks: T100580959

Tags: swift-adoption

Signature: 30983093:1632340205:3bdd3218ae86ad6b3d199cc1b504a625e3650ec0

aschwaighofer reviewed (changes requested):
This is great!

I am only worried about the one case mentioned inline (multi payload enum generation) that we would generate slightly worse code and I think we should address as mentioned inline.

---

While reviewing and playing with generated code I noticed an opportunity for improvement (independent of the code in this PR).


```
public class MyClass {}

public enum SinglepayloadEnum {
    case some(MyClass)
    case none
}

public func copySinglepayloadEnum(_ from: inout SinglepayloadEnum) {
    var to = from
}
```

If you look at the code we generate in the copy value witness for the enum at `-O` vs `-Onone`, surprisingly at `-Onone` the code is arguably be better (at least in terms of code size). That is because we use the TypeInfo based code generation at `-Onone` that has an optimization for this case vs the TypeLayout based generation at `-O` that misses it.

```
define internal %swift.opaque* @"$s1A17SinglepayloadEnumOwcp"(%swift.opaque* noalias %dest, %swift.opaque* noalias %src, %swift.type* %SinglepayloadEnum) #6 {
entry:
  %0 = bitcast %swift.opaque* %dest to %T1A17SinglepayloadEnumO*
  %1 = bitcast %swift.opaque* %src to %T1A17SinglepayloadEnumO*
  %2 = bitcast %T1A17SinglepayloadEnumO* %0 to %swift.refcounted**
  %3 = bitcast %T1A17SinglepayloadEnumO* %1 to %swift.refcounted**
  %4 = load %swift.refcounted*, %swift.refcounted** %3, align 8
  %5 = call %swift.refcounted* @swift_retain(%swift.refcounted* returned %4) #2
  store %swift.refcounted* %4, %swift.refcounted** %2, align 8
  %6 = bitcast %T1A17SinglepayloadEnumO* %0 to %swift.opaque*
  ret %swift.opaque* %6
}

vs (-O)
define internal %swift.opaque* @"$s1A17SinglepayloadEnumOwcp"(%swift.opaque* noalias returned %dest, %swift.opaque* noalias nocapture readonly %src, %swift.type* nocapture readnone %SinglepayloadEnum) #7 {
entry:
  %0 = bitcast %swift.opaque* %src to i64*
  %1 = load i64, i64* %0, align 8
  %2 = and i64 %1, -4294967295
  %3 = icmp ne i64 %2, 0
  %4 = and i64 %1, 8589934590
  %5 = icmp eq i64 %4, 8589934590
  %6 = or i1 %3, %5
  %7 = inttoptr i64 %1 to %T1A7MyClassC*
  br i1 %6, label %payloadBlock, label %noPayloadBlock

payloadBlock:                                     ; preds = %entry
  %8 = bitcast %swift.opaque* %dest to %T1A7MyClassC**
  %9 = getelementptr %T1A7MyClassC, %T1A7MyClassC* %7, i64 0, i32 0
  store %T1A7MyClassC* %7, %T1A7MyClassC** %8, align 8
  %10 = tail call %swift.refcounted* @swift_retain(%swift.refcounted* returned %9) #6
  br label %12

noPayloadBlock:                                   ; preds = %entry
  %11 = bitcast %swift.opaque* %dest to i64*
  store i64 %1, i64* %11, align 8
  br label %12

12:                                               ; preds = %payloadBlock, %noPayloadBlock
  ret %swift.opaque* %dest
}
```

The TypeInfo based generation has a special case that recognizes if you have a single payload enum with one non payload case and just uses a retain/release instead of switching on the payload and then releases. This is valid because the non payload case is going to be represented as `0` which retain/release can handle.

aschwaighofer commented on lib/IRGen/GenEnum.cpp:
```diff
@@ -3509,7 +3509,7 @@ namespace {
       if (AllowFixedLayoutOptimizations && TIK >= Loadable) {
         // The type layout entry code does not handle spare bits atm.
-        return IGM.typeLayoutCache.getOrCreateScalarEntry(getTypeInfo(), T);
+        return IGM.typeLayoutCache.getOrCreateResilientEntry(T);
```
This adds the overhead of loading the value witness in code emission of e.g `copy_addr`. Something like:

```
  copy_addr %3 to [initialization] %2 : $*MultipayloadEnum  
```

Is going to call to the value witness functions (which requires loading the value witness table).


I think, instead we should introduce something like a `TypeInfoBasedTypeLayoutEntry` which is implemented like `ScalarTypeLayoutEntry` before this patch. We can use it until `ScalarTypeLayoutEntry` grows the full ability of TypeInfo based lowering (e.g handle spare bits computations).

gmittert commented:
Added a "TypeInfoBasedTypeLayout" for Layouts that don't have everything implemented yet and moved the CXXStructs and multi enums to use it.

Since we have the information in scalar layouts now, I added a case in the enum destroy to just release the payload if it's a nullable refcounted pointer.

gmittert commented:
@swift-ci please test

gmittert commented:
@swift-ci please smoke test

gmittert commented:
Moving GenStruct to produce Aligned Type layouts inflates code size a ton because it computes an alignment for each field using and/add instead of doing a single gep. Going to split that out into a different PR and try to tackle that there.

gmittert commented:
@swift-ci please test

gmittert commented:
@swift-ci please smoke test

aschwaighofer commented:
@swift-ci Please benchmark

aschwaighofer commented:
@swift-ci Please benchmark

aschwaighofer reviewed (approved):
Lgtm

aschwaighofer commented:
@swift-ci Please smoke test

gmittert commented:
preset=buildbot,tools=RA,stdlib=RD,test=non_executable
@swift-ci Please test with preset macOS
